Add Arducam 64MP driver and device tree overlay

ArduCAM opened the pull request.

Commit:
media: i2c: Add driver of Arducam 64MP camera

Add a driver for the Arducam 64MP camera sensor.
Whilst the sensor supports 2 or 4 CSI2 data lanes, this driver
currently only supports 2 lanes.

The following Bayer modes are currently available:

9152x6944 10-bit @ 2.7fps
4624x3472 10-bit (binned) @ 10fps
3840x2160 10-bit (cropped/binned) @ 20fps
2312x1736 10-bit (binned) @ 30fps
1920x1080 10-bit (cropped/binned) @ 60fps
1280x720 10-bit (cropped/binned) @ 120fps

Signed-off-by: Lee Jackson <info@arducam.com>

Commit:
media: dt-bindings: media: i2c: Add Arducam 64MP CMOS sensor binding

Add YAML device tree binding for Arducam 64MP CMOS image sensor, and
the relevant MAINTAINERS entries.

Signed-off-by: Lee Jackson <info@arducam.com>

Commit:
configs: Add CONFIG_VIDEO_ARDUCAM_64MP=m

Include the driver module for the Arducam 64MP.

Signed-off-by: Lee Jackson <info@arducam.com>

Commit:
overlays: Add arducam-64mp-overlay.dts

Added overlays for enabling Arducam 64MP and add the
relevant information to the README.

Signed-off-by: Lee Jackson <info@arducam.com>

pelwell commented:
Is this your own sensor? If not, it should have the manufacturer's name and model, e.g. "sony,imx477" as used in the driver on which this is based.

And yes, I should have asked the same question of the Pivariety driver.

6by9 commented on Documentation/devicetree/bindings/media/i2c/arducam,64mp.yaml:
```diff
@@ -0,0 +1,113 @@
+              For two-lane operation the property must be set to <1 2>.
+            items:
+              - const: 1
+              - const: 2
```
The sensor supports either 2 or 4 lanes, so the binding should follow suit and list both options.

ArduCAM replied:
Added a new items, a little unsure, since the camera driver does not have 4lane configuration, will this affect dt-bindings?

6by9 commented on drivers/media/i2c/arducam_64mp.c:
```diff
@@ -0,0 +1,2630 @@
+	{0x3092, 0x01},
+	{0x3093, 0x00},
+	{0x0350, 0x00},
+};
```
Just doing a simple diff on these tables seems to imply that there are about 50% of them that don't change between modes. Could those be moved into ```mode_common_regs```?

ArduCAM replied:
I moved some registers into common.

6by9 commented on drivers/media/i2c/arducam_64mp.c:
```diff
@@ -0,0 +1,2630 @@
+			.width = 9248,
+			.height = 6944,
+		},
+		.timeperframe_min = {
```
```timeperframe_min ``` is the same as ```timeperframe_default ``` in each of these cases. Do we need both?

ArduCAM replied:
timeperframe_min is removed, treat the default as the maximum achievable framerate.

6by9 commented on drivers/media/i2c/arducam_64mp.c:
```diff
@@ -0,0 +1,2630 @@
+ * in the datasheet) is much smaller - 600us.
+ */
+#define ARDUCAM_64MP_XCLR_MIN_DELAY_US	8000
+#define ARDUCAM_64MP_XCLR_DELAY_RANGE_US	1000
```
This text and the numbers are identical to imx477. Is it exactly the same (including the timings being T6, T7, and T8), or is this just copy/paste?

ArduCAM replied:
Missed it out, it's a little different from imx477 and has been modified.

6by9 commented on drivers/media/i2c/arducam_64mp.c:
```diff
@@ -0,0 +1,2630 @@
+};
+
+/* Mode configs */
+static const struct arducam_64mp_mode supported_modes_10bit[] = {
```
You appear to only have 10bit modes, so this can be ```supported_modes``` without reference to the bit depth.

ArduCAM replied:
done.

6by9 commented on drivers/media/i2c/arducam_64mp.c:
```diff
@@ -0,0 +1,2630 @@
+	struct arducam_64mp *arducam_64mp =
+		container_of(ctrl->handler, struct arducam_64mp, ctrl_handler);
+	struct i2c_client *client = v4l2_get_subdevdata(&arducam_64mp->sd);
+	int ret = 0;
```
I don't see a path through the function which doesn't set ```ret```, so no need to initialise to 0.

ArduCAM replied:
done.

6by9 commented on drivers/media/i2c/arducam_64mp.c:
```diff
@@ -0,0 +1,2630 @@
+
+	/*
+	 * The sensor must be powered for arducam_64mp_identify_module()
+	 * to be able to read the CHIP_ID register
```
You're not reading the CHIP_ID register of the sensor, so does this apply?

Checking on my preproduction module, the EEPROM does appear to be controlled by the same regulator, so presumably yes it is required. The comment could be updated though.

ArduCAM replied:
You are right, I simply modified the comment.

6by9 commented on arch/arm/boot/dts/overlays/arducam-64mp-overlay.dts:
```diff
@@ -0,0 +1,94 @@
+							/bits/ 64 <456000000>;
+					};
+				};
+			};
```
The module includes a VCM for focus control. Presumably there will be an follow up update to add in support for that.

6by9 commented on Documentation/devicetree/bindings/media/i2c/arducam,64mp.yaml:
```diff
@@ -0,0 +1,113 @@
+  The Arducam 1/1.7-Inch 64Mpixel CMOS active pixel digital image sensor
+  with an active array size of 9248 x 6944. It is programmable through
+  I2C interface. The I2C address is fixed to 0x1A as per sensor data sheet.
+  Image data is sent through MIPI CSI-2, which is configured as either 2 or
```
which can be configured for operation with either 2 or 4 data lanes.

ArduCAM replied:
done.

6by9 commented on Documentation/devicetree/bindings/media/i2c/arducam,64mp.yaml:
```diff
@@ -0,0 +1,113 @@
+            type: boolean
+            description: |-
+              MIPI CSI-2 clock is non-continuous if this property is present,
+              otherwise it's continuous.
```
I don't see usage of this flag in the driver.

Checking the register set against the data sheet the driver appears to only ever run in non-continuous mode (the default). In which case the driver ought to ensure that flag is set.

Then again the joys of V4L2 means that you can still incorrectly configure the CSI2 receiver to continuous mode regardless of what the sensor driver has been told.

ArduCAM replied:
I'm not sure if my change is correct, I simply changed it to `clock-noncontinuous: true`

6by9 reviewed (commented):
Generally minor comments.

Use of a second pad for metadata isn't acceptable for mainline, but is how embedded data is handled currently on the Pi branches. Assuming it actually works on this sensor, then that is fine by me.

Commit:
media: i2c: arducam_64mp: Advertise embedded data node on media pad 1

This commit updates the arducam_64mp driver to adverise support for
embedded data streams.

The arducam_64mp sensor subdevice overloads the media pad to differentiate
between image stream (pad 0) and embedded data stream (pad 1) when
performing the v4l2_subdev_pad_ops functions.

Signed-off-by: Lee Jackson <info@arducam.com>

ArduCAM commented:
> Use of a second pad for metadata isn't acceptable for mainline, but is how embedded data is handled currently on the Pi branches. Assuming it actually works on this sensor, then that is fine by me.

Yes, it works. I missed this, I think I'd better put the embedded part of the code into a separate commit.

pelwell commented:
Do you have an answer for my question?

> Is this your own sensor? If not, it should have the manufacturer's name and model, e.g. "sony,imx477" as used in the driver on which this is based.

> And yes, I should have asked the same question of the Pivariety driver.

ArduCAM commented:
Pelwell
Sorry, I intended to reply you through email phil.elwell, but seems it doesn't work.
You might not know that the idea of Pivariety solution is making RPI camera various and one driver for all cameras, user doesn't need to switch back and forth with dt-overlay settings when switching different cameras from us. We have a huge camera lists in mind including global shutter camera, ToF camera, Event-based camera, dual-lenses camera with extended features like external trigger, Strobe output, etc. Although these sensors are not our own, we would like to make them uniformed to kernel driver side in our humble opinion.

pelwell commented:
My aim is to avoid having a number of different drivers for the same sensor packaged by different manufactures - it's better to have one generic driver - hence my question about the sensor. Feel free to continue the conversation by email if you prefer - I'm just `phil@raspberrypi.com`.

adminy commented:
Can this driver be upstreamed to the mainline linux kernel?
